backend/cmm_helpers.ml

```diff
@@ -38,6 +38,7 @@ let bind_nonvar name arg fn =
   | _ -> let id = V.create_local name in Clet(VP.create id, arg, fn (Cvar id))
 
 let caml_black = Nativeint.shift_left (Nativeint.of_int 3) 8
+let caml_local = Nativeint.shift_left (Nativeint.of_int 2) 8
```

Comment:
Could you explain why you're using the blue colour instead of black ? Is it related to the colour changes in multicore ?

Reply:
Local blocks use blue so that the GC can efficiently distinguish which roots point to local allocations. (I hadn't realised this GC interaction would appear in this PR, but of course the backend builds headers directly).

This is unrelated to the multicore use of colours.

Reply:
Hmm... I'm not sure why the GC needs to know that. For instrumentation or debugging purposes, maybe ? Unless you're not adding the local stack as roots and rely on the variables on the normal stack pointing to local allocations to know which ones are live.

Reply:
> Unless you're not adding the local stack as roots and rely on the variables on the normal stack pointing to local allocations to know which ones are live.

Yes, exactly this. (Not necessarily on the normal stack; registers too)

Reply:
I think the problem is the following: since the gc does not scan (and particularly, does not follow pointers inside of) a block with black headers, if a heap block is only kept alive by a local allocation that has a black header, then the GC might incorrectly consider the heap block dead ?

Reply:
That was my point: you could consider all values on the local stack as roots, like the global symbols (which can be roots even though they have black headers). I'm not sure how complicated it would be (you don't have an easy way to iterate on the headers, so you would have to hope that the stack is full of well-formed blocks with no empty space in-between), but since that would have been my first idea I wonder how well it would fare in practice.

Reply:
We could indeed consider all values on the local stack as roots, and we've ensured that the local stack consists only of well-formed blocks with no empty space so it can be traversed. This is actually how the earliest implementation worked.

However, we're doing inference of allocation modes, so heap blocks can silently be made local when possible. Currently, in the following computation:
```
let f a b =
  let pair = (a, b) in
  let x = fast_computation pair in
  let y = slow_computation () in
  (x, y)
```
the block `pair` as well as the values `a` and `b` are eligible for GC during `slow_computation`. If we silently made `pair` local (e.g. because fast_computation allows local arguments), then scanning all of the locals would worsen the memory usage of this program by making `a` and `b` seem alive during `slow_computation`, because `pair` is not popped off the local stack until the function returns.

So instead, the root-scanning logic scans only live local values, and uses header colour to know which ones those are. `pair` will still only be popped when the function returns, but the values `a` and `b` remain eligible for GC during `slow_computation`.